pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,203 @@
+package leaderelection
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+
+	"github.com/razorpay/metro/pkg/logger"
+
+	"github.com/razorpay/metro/pkg/registry"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	// JitterFactor used in wait utils
+	JitterFactor = 1.2
+)
+
+// NewLeaderElector creates a LeaderElector from a LeaderElection Config
+func NewLeaderElector(c Config, registry registry.Registry) (*LeaderElector, error) {
+	if err := c.Validate(); err != nil {
+		return nil, err
+	}
+
+	le := LeaderElector{
+		config:   c,
+		registry: registry,
+		nodeID:   "",
+	}
+
+	return &le, nil
+}
+
+// LeaderElector is a leader election client.
+type LeaderElector struct {
+	// nodeID represents the id assigned by registry
+	nodeID string
+
+	// leaderID stores id the current leader
+	leaderID string
+
+	// Leader election config
+	config Config
+
+	// registry being used for leader election
+	registry registry.Registry
+}
+
+// Run starts the leader election loop. Run will not return
+// before leader election loop is stopped by ctx or it has
+// stopped holding the leader lease
+func (le *LeaderElector) Run(ctx context.Context) {
+	defer runtime.HandleCrash()
```

Comment:
HandleCrash() does `recover()` which might be riskier, go doesn't encourage recovering. Lets remove this for now

Reply:
will remove